### PR TITLE
pr: 19 pdbert 데모 실행 점검 스크립트 작성

### PR DESCRIPTION
**Pull request by MJ-bin**
# 주요 변경점
### pdbert 를 위한 test_image.bats 스크립트를 작성했습니다.
* MJbin 유저의 PDBERT 레포에 추가된 [prepare_dataset.py](https://github.com/MJ-bin/PDBERT/blob/main/prepare_dataset.py) 을 이용하게 로직을 수정했습니다.

### linevul 을 위한 test_image.bats 스크립트를 수정했습니다.
* 기존의 파일 존재 여부만 검증하는 로직에서, 터미널에 출력되는 문자열을 매칭하여 pass 하게끔 수정했습니다.

### Readme.md
* Readme 파일에 pdbert 를 위한 bats 스크립트 사용법을 추가하였습니다.

### CODE REVIEW

**Review by seokjeon — COMMENTED** [review]
끝나면 다시 리뷰 요청 바랍니다.

**Comment by seokjeon** [comment]
#19에서 가져온 내용

linevul 검증 결과
$ bats ./docker/linevul/test_image.bats
test_image.bats
   Using existing compose container: linevul
 ✓ 1. 불완전한 jasper 데이터셋에 processed_func 열 추가
   append_datasetjasper.py output:
Loading source code mapping...
  - Indexed 270917 source files

Processing Real_Vul_data.csv (filtering jasper & merging source)...

Completed:
  - Jasper rows processed: 315
  - Source merged: 315
  - Missing source: 0
  - Output: /app/RealVul/Dataset/jasper_data_append_processed_func.csv
   Dataset append completed successfully
   2. Jasper 데이터셋의 pickle 생성 (--prepare_dataset)                       2/4
   prepare_dataset output:
2026-01-04 09:52:36,295 INFO services.py:2021 -- object_store_memory is not verified when plasma_directory is set.
 ✓ 2. Jasper 데이터셋의 pickle 생성 (--prepare_dataset)
Arguments Namespace(dataset_csv_path='/app/RealVul/Dataset/jasper_data_append_processed_func.csv', dataset_path='/app/RealVul/Dataset/', model_name='/app/RealVul/Experiments/LineVul/best_model', num_train_epochs=10, output_dir='/app/RealVul/Experiments/LineVul', per_device_eval_batch_size=8, per_device_train_batch_size=8, prepare_dataset=True, test_predict=False, tokenizer_name='microsoft/codebert-base', train=False, train_predict=False, val_predict=False)
Preparing Dataset...
 38%|███▊      | 70/183 [00:02<00:04, 23.94it/s](read_file_label pid=86025) Token indices sequence length is longer than the specified maximum sequence length for this model (3225 > 512). Running this sequence through the model will result in indexing errors
100%|██████████| 183/183 [00:06<00:00, 28.44it/s]
100%|██████████| 687/687 [00:00<00:00, 115075.35it/s]
 15%|█▌        | 20/132 [00:00<00:03, 33.92it/s](read_file_label pid=86067) Token indices sequence length is longer than the specified maximum sequence length for this model (3225 > 512). Running this sequence through the model will result in indexing errors [repeated 64x across cluster] (Ray deduplicates logs by default. Set RAY_DEDUP_LOGS=0 to disable log deduplication, or see https://docs.ray.io/en/master/ray-observability/ray-logging.html#log-deduplication for more options.)
100%|██████████| 132/132 [00:03<00:00, 33.75it/s]
100%|██████████| 604/604 [00:00<00:00, 88743.46it/s]
(read_file_label pid=86019) Token indices sequence length is longer than the specified maximum sequence length for this model (8106 > 512). Running this sequence through the model will result in indexing errors [repeated 61x across cluster]
   Dataset preparation completed successfully
   3. Jasper 데이터셋으로 학습 (--train)                                      3/4
   train output:
2026-01-04 09:52:55,997 INFO services.py:2021 -- object_store_memory is not verified when plasma_directory is set.
2026-01-04 09:52:56,137 INFO worker.py:1724 -- Started a local Ray instance.
/usr/local/lib/python3.8/dist-packages/transformers/optimization.py:306: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(
***** Running training *****
  Num examples = 605
  Num Epochs = 10
  Instantaneous batch size per device = 8
  Total train batch size (w. parallel, distributed & accumulation) = 8
  Gradient Accumulation steps = 1
  Total optimization steps = 760
  Number of trainable parameters = 124647170
Arguments Namespace(dataset_csv_path='/app/RealVul/Dataset/jasper_data_append_processed_func.csv', dataset_path='/app/RealVul/Dataset/', model_name='/app/RealVul/Experiments/LineVul/best_model', num_train_epochs=10, output_dir='/app/RealVul/Experiments/LineVul', per_device_eval_batch_size=8, per_device_train_batch_size=8, prepare_dataset=False, test_predict=False, tokenizer_name='microsoft/codebert-base', train=True, train_predict=False, val_predict=False)
Loading Dataset...
Training Dataset...
 10%|█         | 76/760 [02:15<16:30,  1.45s/it]***** Running Evaluation *****
  Num examples = 604
  Batch size = 8
 10%|█         | 76/760 [02:47<16:30,  1.45s/itSaving model checkpoint to /app/RealVul/Experiments/LineVul/checkpoint-76         
Configuration saved in /app/RealVul/Experiments/LineVul/checkpoint-76/config.json
Model weights saved in /app/RealVul/Experiments/LineVul/checkpoint-76/pytorch_model.bin
 20%|██        | 152/760 [04:55<16:09,  1.59s/it]***** Running Evaluation *****
  Num examples = 604
  Batch size = 8
                                      {'eval_loss': 0.13976263999938965, 'eval_accuracy': 0.9685430463576159, 'eval_precision': 0.59375, 'eval_recall': 0.76, 'eval_f1': 0.6666666666666666, 'eval_confusion_matrix': [[566, 13], [6, 19]], 'eval_runtime': 32.4775, 'eval_samples_per_second': 18.597, 'eval_steps_per_second': 2.34, 'epoch': 1.0}
 20%|██        | 152/760 [05:28<16:09,  1.59s/iSaving model checkpoint to /app/RealVul/Experiments/LineVul/checkpoint-152        
Configuration saved in /app/RealVul/Experiments/LineVul/checkpoint-152/config.json
Model weights saved in /app/RealVul/Experiments/LineVul/checkpoint-152/pytorch_model.bin
 30%|███       | 228/760 [07:38<12:56,  1.46s/it]***** Running Evaluation *****
  Num examples = 604
  Batch size = 8
                                      {'eval_loss': 0.22082391381263733, 'eval_accuracy': 0.9586092715231788, 'eval_precision': 0.5, 'eval_recall': 0.48, 'eval_f1': 0.4897959183673469, 'eval_confusion_matrix': [[567, 12], [13, 12]], 'eval_runtime': 33.0809, 'eval_samples_per_second': 18.258, 'eval_steps_per_second': 2.297, 'epoch': 2.0}
 30%|███       | 228/760 [08:10<12:56,  1.46s/iSaving model checkpoint to /app/RealVul/Experiments/LineVul/checkpoint-228        
Configuration saved in /app/RealVul/Experiments/LineVul/checkpoint-228/config.json
Model weights saved in /app/RealVul/Experiments/LineVul/checkpoint-228/pytorch_model.bin
 40%|████      | 304/760 [10:15<10:51,  1.43s/it]***** Running Evaluation *****
  Num examples = 604
  Batch size = 8
{'eval_loss': 0.1960097998380661, 'eval_accuracy': 0.9619205298013245, 'eval_precision': 0.5384615384615384, 'eval_recall': 0.56, 'eval_f1': 0.5490196078431373, 'ev ✓ 3. Jasper 데이터셋으로 학습 (--train)
s_per_second': 18.883, 'eval_steps_per_second': 2.376, 'epoch': 3.0}
 40%|████      | 304/760 [10:47<10:51,  1.43s/iSaving model checkpoint to /app/RealVul/Experiments/LineVul/checkpoint-304        
Configuration saved in /app/RealVul/Experiments/LineVul/checkpoint-304/config.json
Model weights saved in /app/RealVul/Experiments/LineVul/checkpoint-304/pytorch_model.bin


Training completed. Do not forget to share your model on huggingface.co/models =)


Loading best model from /app/RealVul/Experiments/LineVul/checkpoint-76 (score: 0.13976263999938965).
 40%|████      | 304/760 [10:50<16:16,  2.14s/it]
Saving model checkpoint to /app/RealVul/Experiments/LineVul/best_model
Configuration saved in /app/RealVul/Experiments/LineVul/best_model/config.json
Model weights saved in /app/RealVul/Experiments/LineVul/best_model/pytorch_model.bin
{'eval_loss': 0.18353497982025146, 'eval_accuracy': 0.9619205298013245, 'eval_prec ✓ 4. Jasper 데이터셋으로 테스트 (--test_predict)
val_confusion_matrix': [[567, 12], [11, 14]], 'eval_runtime': 31.4866, 'eval_samples_per_second': 19.183, 'eval_steps_per_second': 2.414, 'epoch': 4.0}
{'train_runtime': 650.7744, 'train_samples_per_second': 9.297, 'train_steps_per_second': 1.168, 'train_loss': 0.10554136727985583, 'epoch': 4.0}
   Training completed successfully
   4. Jasper 데이터셋으로 테스트 (--test_predict)                             4/4
   test_predict output:
2026-01-04 10:03:56,252 INFO services.py:2021 -- object_store_memory is not verified when plasma_directory is set.
2026-01-04 10:03:57,391 INFO worker.py:1724 -- Started a local Ray instance.
***** Running Prediction *****
  Num examples = 604
  Batch size = 8
Arguments Namespace(dataset_csv_path='/app/RealVul/Dataset/jasper_data_append_processed_func.csv', dataset_path='/app/RealVul/Dataset/', model_name='/app/RealVul/Experiments/LineVul/best_model', num_train_epochs=10, output_dir='/app/RealVul/Experiments/LineVul', per_device_eval_batch_size=8, per_device_train_batch_size=8, prepare_dataset=False, test_predict=True, tokenizer_name='microsoft/codebert-base', train=False, train_predict=False, val_predict=False)
Loading Dataset...
Test Results...
100%|██████████| 76/76 [01:24<00:00,  1.11s/it]
Test Metrics {'accuracy': 0.9685430463576159, 'precision': 0.59375, 'recall': 0.76, 'f1': 0.6666666666666666, 'confusion_matrix': [[566, 13], [6, 19]]}
   Test prediction completed successfully

4 tests, 0 failures
pdbert 검증 결과
root@da44f36edc08:/PDBERT# bats ./test_image.bats
    prepare_dataset.py output: pdbert                                              1/2 
   JSON files created successfully
 ✓ 1. jasper 데이터셋 생성 for pdbert
   2. PDBERT 학습 및 평가 (Jasper 데이터셋)                                        2/2 
   PDBERT training output:
   Training and evaluation completed successfully
 ✓ 2. PDBERT 학습 및 평가 (Jasper 데이터셋)

2 tests, 0 failures
pdbert 의 경우 test 도중 에러가 발생하면, 상세한 로그가 표시됩니다.

참고: pdbert는 h100 에서 학습되지 않아서, 수동으로 컨테이너 내부로 cp한후 실행한 결과입니다.

**Comment by seokjeon** [comment]
마찬가지로 윈도우 컴퓨터에서 실행 점검을 어떻게 하면 좋을지 고민해봅시다...

**Comment by MJ-bin** [comment]
### wsl 을 이용한 pdbert bats 스크립트 실행 성공

```bash
mjbin@DESKTOP-BQI8RT3:/mnt/d/RealVul_prj/VP-Bench/docker/pdbert$ bats test_image.bats
   Using existing compose container: pdbert
    prepare_dataset.py output: pdbert                                                                               1/2
   JSON files created successfully
 ✓ 1. jasper 데이터셋 생성 for pdbert
    PDBERT training output:Jasper 데이터셋)                                                                         2/2
   Training and evaluation completed successfully
 ✓ 2. PDBERT 학습 및 평가 (Jasper 데이터셋)

2 tests, 0 failures
```
